src/com/facebook/buck/swift/SwiftDescriptions.java

```diff
@@ -16,60 +16,97 @@
 
 package com.facebook.buck.swift;
 
+import static com.facebook.buck.swift.SwiftUtil.Constants.SWIFT_EXTENSION;
+
 import com.facebook.buck.apple.AppleCxxPlatform;
 import com.facebook.buck.apple.ApplePlatforms;
 import com.facebook.buck.apple.MultiarchFileInfo;
+import com.facebook.buck.cxx.CxxConstructorArg;
 import com.facebook.buck.cxx.CxxPlatform;
+import com.facebook.buck.model.BuildTarget;
 import com.facebook.buck.model.BuildTargets;
+import com.facebook.buck.model.Flavor;
 import com.facebook.buck.model.FlavorDomain;
 import com.facebook.buck.rules.BuildRule;
 import com.facebook.buck.rules.BuildRuleParams;
 import com.facebook.buck.rules.BuildRuleResolver;
 import com.facebook.buck.rules.SourcePath;
 import com.facebook.buck.rules.SourcePathResolver;
-import com.facebook.buck.rules.TargetGraph;
+import com.facebook.buck.rules.SourceWithFlags;
 import com.facebook.buck.rules.Tool;
-import com.facebook.buck.rules.coercer.FrameworkPath;
 import com.facebook.buck.util.HumanReadableException;
+import com.facebook.buck.util.MoreStrings;
 import com.google.common.base.Optional;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSortedSet;
 
 public class SwiftDescriptions {
-  /** Utility class: do not instantiate. */
-  private SwiftDescriptions() {}
+  /**
+   * Utility class: do not instantiate.
+   */
+  private SwiftDescriptions() {
+  }
 
-  @SuppressWarnings("unused")
-  static BuildRule createSwiftModule(
-      FlavorDomain<CxxPlatform> cxxPlatformFlavorDomain,
+  public static <T extends CxxConstructorArg> Optional<BuildRule> generateCompanionSwiftBuildRule(
+      BuildRuleParams parentParams,
+      BuildRuleResolver buildRuleResolver,
+      T args,
       CxxPlatform defaultCxxPlatform,
-      FlavorDomain<AppleCxxPlatform> appleCxxPlatformFlavorDomain,
-      TargetGraph targetGraph,
-      BuildRuleParams params,
-      BuildRuleResolver resolver,
-      CxxPlatform platform,
-      String moduleName,
-      ImmutableSortedSet<SourcePath> sources,
-      ImmutableList<String> compilerFlags,
-      ImmutableSortedSet<FrameworkPath> frameworks,
-      ImmutableSortedSet<FrameworkPath> libraries,
-      boolean enableObjcInterop) {
+      FlavorDomain<CxxPlatform> cxxPlatformFlavorDomain,
+      FlavorDomain<AppleCxxPlatform> platformFlavorsToAppleCxxPlatforms) {
+    BuildTarget parentTarget = parentParams.getBuildTarget();
+    CxxPlatform cxxPlatform = cxxPlatformFlavorDomain.getValue(parentTarget)
+        .or(defaultCxxPlatform);
+    // TODO(tho@uber.com) support version in swift flavor.
+    BuildTarget swiftCompanionTarget = parentTarget.withFlavors(
```

Comment:
I did that on purpose since I didn't want compilation step for swift to run twice (first for preprocess to generate the header, at this step, most of flavors from it parent are not yet passed through, second is the real compile step for building the module). We could end up with two generated headers for example: `App1Binary#swift/App1Binary-Swift.h` and `App1Binary#dwarf-and-dsym,iphonesimulator-x86_64,strip-non-global,swift/App1Binary-Swift.h`, but I think you are right, it is more importing to pass the flavors through.